pinot-core/src/main/java/org/apache/pinot/core/transport/AsyncQueryResponse.java

```diff
@@ -134,7 +148,15 @@ void markRequestSent(ServerRoutingInstance serverRoutingInstance, int requestSen
 
   void receiveDataTable(ServerRoutingInstance serverRoutingInstance, DataTable dataTable, int responseSize,
       int deserializationTimeMs) {
-    _responseMap.get(serverRoutingInstance).receiveDataTable(dataTable, responseSize, deserializationTimeMs);
+    ServerResponse response = _responseMap.get(serverRoutingInstance);
+    response.receiveDataTable(dataTable, responseSize, deserializationTimeMs);
+
+    // Record query completion stats immediately after receiving the response from the server instead of waiting
```

Comment:
Not sure I follow this change. Was this a bug ?

Reply:
Not a bug. I just consolidated all `recordStatsUponResponseArrival` into AsyncResponse instead of keeping it scattered all over.  It makes sense to be here because if our response handling changes in the future, the author will be reminded to handle it.

Reply:
Got it. Thanks for clarifying.